### DISCUSSION TITLE
Language Server Protocol

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#408] [Ideas]
Pour avoir une meilleure expérience de développement d'un fichier Publicode, il faudrait interfacer le “moteur” avec le [Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol).

Parmi les fonctionnalités qui seraient les plus utiles :
* Navigation facilitée dans les règles (<kbd>F12</kbd> “Go to definition”)
* Renommage d'une règle qui modifie toutes les références avec une compréhension sémantique des espaces de noms (<kbd>F2</kbd> pour renommer) 

Un des aspects à regarder, c'est notamment que le développement d'un “language server” nécessite d'avoir un parseur tolérant aux erreurs. Microsoft a publié un retour d'expérience intéressant pour l’intégration d'une telle logique dans le compilateur PHP : https://github.com/Microsoft/tolerant-php-parser/blob/master/docs/HowItWorks.md

Pour avoir un bon niveau d'intégration il faudra aussi probablement utiliser une API de plus bas niveau pour le parsage du `yaml` afin d'avoir une bonne intégration des tokens tokens et AST dans une structure unique pour les fichiers publicode. La libraire NPM `yaml` que nous utilisions propose ce type d'API de bas niveau https://eemeli.org/yaml/#api-overview

Je ne suis pas certain que nous ayons la bande passante mener à bien ces travaux dès maintenant, mais c'est un besoin à garder en tête notamment si on ré-écrit des parties du parseur.

Références :
*  https://microsoft.github.io/language-server-protocol/
* https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Comment by laem** [comment]
@ivangabriele

**Comment by guillett** [comment]
On pourrait faire la même chose sur OpenFisca ? :thinking:  (Un peu :trollface: mais pas que ! )

**Comment by johangirod** [comment]
Voir le travail d'@EmileRolley
https://github.com/EmileRolley/publicodes-language-server

**Comment by EmileRolley** [comment]
Le code est déplacé dans l'orga à présent : https://github.com/publicodes/language-server

Les prochains sujets en lien sur lesquels je vais investiguer dans les prochaines semaines :
1. Rédaction d'une grammaire en ENBF pour publicodes (en incluant une forme simplifié de YAML dans le language)
2. Implémentation d'un parseur [`tree-sitter`](https://tree-sitter.github.io/tree-sitter/creating-parsers#writing-the-grammar)
3. (Réécrire le serveur à partir de ce parseur)

**Comment by EmileRolley** [comment]
Je pense que l'on peut fermer cette discussion à présent. Les issues et questions relatives au LS devront être traitées directement dans le dépôt dédié : https://github.com/publicodes/language-server.
